### ISSUE TITLE
xfce: When caja is run, it requires 6quits to end the program

### ISSUE THREAD

**Issue by raveit65** [mate-desktop/caja#481]
origin upstream report:
https://bugzilla.redhat.com/show_bug.cgi?id=1278079
```
 John Duchek 2015-11-04 11:35:17 EST

Description of problem:

If I run caja, it display a directory. I click on the upper right x to close it, and it reopens. If I reclick the x, it closes and reopens.  This occurs for a total of 6 times until the program closes for good.
Version-Release number of selected component (if applicable):
1.10.4

How reproducible:
Every time it runs

Steps to Reproduce:
1. Click on caja
2.click x to close
3.repeat

Actual results:
reopens 6 times

Expected results:
Close the first time I tell it to.


Additional info:
I am running F23 xfce window manager
```
```
 Kevin Fenzi 2015-11-08 11:29:35 EST

I can confirm the behavior, but not sure why it's doing it. ;( 

I do see:

(caja:30907): Eel-WARNING **: "unique eel_ref_str" hash table still has 4 elements at quit time (keys above)
(caja:30907): Eel-WARNING **: "caja-directory.c: directories" hash table still has 2 elements at quit time

Perhaps that means something to the mate folks?
```
```
 Wolfgang Ulbrich 2015-11-08 14:12:54 EST

(In reply to Kevin Fenzi from comment #4)
> I can confirm the behavior, but not sure why it's doing it. ;( 
> 
> I do see:
> 
> (caja:30907): Eel-WARNING **: "unique eel_ref_str" hash table still has 4
> elements at quit time (keys above)
> (caja:30907): Eel-WARNING **: "caja-directory.c: directories" hash table
> still has 2 elements at quit time
> 
> Perhaps that means something to the mate folks?

What you see here is the result of an improvement by this commit.
https://github.com/mate-desktop/caja/commit/4b0d231fdf8dfd6d299ec07cc933000a610f0a74
This avoid that caja hijack the desktop in another enviroment or if you start caja as root.
Normaly caja handles the desktop in mate-session, ie. icons and contextmenus.
This warnings i see in mate-session too if i start caja as root.
So, it's a bit unusual to print such warnings (debuginfos) in a terminal but it's nothing bad and never cause this issue in xfce seesion, imo.
And those are only warnings.
```
```
 Kevin Fenzi 2015-11-08 14:43:43 EST

The logic there seems a bit wrong to me. 

It says:

if you are running as root and not on mate, exit when the window closes. 
If you are not running as root and not on mate, get the value from gsettings. 

This fails for any user that hasn't set the gsetting value. (Like most Xfce users)

Indeed, if I run caja as root under Xfce it correctly exits on the first close. 

So, I think line 525 there should be shortened to: 

if (running_in_mate ())

Or perhaps it should check if that pref is set no matter what desktop and if it's not set default it to true.
```

**Comment by nirik** [comment]
I'm one of the downstream folks that see this. Happy to test or provide additional information. Thanks.

**Comment by agp8x** [comment]
I'm also affect by this (Xubuntu 15.10 with MATE installed)
> --- Hash table keys for warning below:
> --> l2081
> --> agp8x
> --> inode/directory
> 
> (caja:580): Eel-WARNING **: "unique eel_ref_str" hash table still has 3 elements at quit time (keys above)
> 
> (caja:580): Eel-WARNING **: "caja-directory.c: directories" hash table still has 2 elements at quit time

**Comment by raveit65** [comment]
@monsta 
Can you please take a look into it?

**Comment by monsta** [comment]
Ok, reproduced in Xubuntu 15.10.

BTW, https://bugzilla.redhat.com/1250404 looks similar.

**Comment by monsta** [comment]
However, I can't reproduce it in Cinnamon, so I doubt the code logic mentioned above is the cause...

**Comment by monsta** [comment]
>  Kevin Fenzi 2015-11-08 14:43:43 EST

> It says:

> if you are running as root and not on mate, exit when the window closes. 
If you are not running as root and not on mate, get the value from gsettings. 

> This fails for any user that hasn't set the gsetting value. (Like most Xfce users)

Let's look at the code:

```C
/* exit_with_last_window is already set to TRUE, and we need to keep that value
 * on other desktops or when running caja as root. Otherwise, we read the value
 * from the configuration.
 */
if (running_in_mate () && !running_as_root())
{
    exit_with_last_window = g_settings_get_boolean (caja_preferences, CAJA_PREFERENCES_EXIT_WITH_LAST_WINDOW);
}
```
So:
- "if you are running as root and not on mate, exit when the window closes" - yes.
- "If you are not running as root and not on mate, get the value from gsettings" - no. The ```running_in_mate ()``` check fails in *both* cases, so ```exit_with_last_window``` settings stays enabled, and no reading from gsettings takes place.

And I see that Caja correctly processes this setting under Xfce. It indeed exits with the last window closed. If it wasn't so, there would be a background Caja process still running after you close the last window - but then Xfce session would see that Caja process is still running, and wouldn't try to restart Caja at all. :smile:

> Indeed, if I run caja as root under Xfce it correctly exits on the first close. 

It might simply be that Xfce session doesn't take care of processes running as root.
Why not? It wouldn't have the rights to auto-restart them as root again anyway. :smile:

**Comment by nirik** [comment]
Xfce doesn't do any "auto restarting" of apps. 

If this isnt the problem area then I am not sure what is. 

I do see that the version of caja in Fedora is 1.12.1. Perhaps thats using an older code base with different issues?

**Comment by raveit65** [comment]
> However, I can't reproduce it in Cinnamon, so I doubt the code logic mentioned above is the cause...

cinnamon use dconf per default, i'm not sure if xfce is able to read/follow dconf settings in general

**Comment by raveit65** [comment]
> BTW, https://bugzilla.redhat.com/1250404 looks similar.
f21 is eol and this report was with caja 1.8.x version w/o the fix we are talking about.
....indeed it looks like the same causes

**Comment by raveit65** [comment]
@nirik 
in regards to the olderrhbz report
Does 'caja --sm-client-disable --no-desktop --browser' help?
If yes, --sm-client-disable does the magic

**Comment by nirik** [comment]
Xfce doesn't use dconf or gconf, it has it's own xfconf. ;) 

Yes, the --sm-client-disable makes it exit the first time. 

Without that it takes closing it exactly 6 times to make it exit.

**Comment by raveit65** [comment]
> Xfce doesn't use dconf or gconf, it has it's own xfconf. ;)
Well, i know that.
> Yes, the --sm-client-disable makes it exit the first time. 
In other words xfce session management restart caja, if no other session manager is running.
It looks like that https://github.com/mate-desktop/caja/blob/master/data/caja.desktop.in.in#L17
is read by xfce session management :-/
I have no clue how to surpress this for xfce.

**Comment by nirik** [comment]
> In other words xfce session management restart caja, if no other session manager is running.
It looks like that https://github.com/mate-desktop/caja/blob/master/data/caja.desktop.in.in#L17
is read by xfce session management :-/
I have no clue how to surpress this for xfce.

No. Xfce doesn't read or care about any X-MATE entries in desktop files. If Xfce was restarting it, why would it do so exactly 6 times and then exit?

**Comment by raveit65** [comment]
--sm-client-disable means that caja starts w/o control by mate-session- manager.
If i use this option in mate-session + a killall, caja doesn't restart and all desktop icons are gone.

**Comment by monsta** [comment]
> Xfce doesn't do any "auto restarting" of apps. 

It doesn't?

![xfce4-session-settings](https://cloud.githubusercontent.com/assets/5138986/11686788/0dbeb42e-9e94-11e5-9ff3-956e3a27a96c.png)

Ok, then what "Restart style" column is for?

And why Caja quits on the first try if I set "Restart Style" to "Never" while Caja is running? It works (until you manually start Caja next time though). Try it.

**Comment by raveit65** [comment]
If i edit /usr/share/applications/caja-desktop
X-MATE-Autostart-Notify=true
to
X-MATE-Autostart-Notify=false

restart style is set to'if running' in xfce4-session-settings and caja close cleanly.
@nirik 
sorry, but it looks like that xfce session management read this value.
Anyways, we need to find out why this happens.

**Comment by raveit65** [comment]
Using
X-GNOME-AutoRestart=true
works too and caja closed cleanly, but this breaks cajas behaviour in mate-session.

**Comment by nirik** [comment]
ok. I finally had some time to dig more into this and you are correct that it is Xfce restarting it. ;) Appologies.

What I think is happening is this: 

caja starts, parses it's desktop file and sees the X-MATE-AutoRestart=true and sets:

                xsmp->restart_style = SmRestartImmediately;

Xfce picks this up as a xmp session hint and keeps restarting it. 

This code seems to be in:

caja-1.12.1/cut-n-paste-code/libegg/eggsmclient-xsmp.c

I'm not clear how caja links or uses that code, but it seems to be where this is happening. 
So, perhaps a check could be added there to not set that if not running in MATE?

Thoughts?

**Comment by monsta** [comment]
Indeed... that's kind of surprise to me :smile:
The code is used when communicating with session manager... honestly, I thought it's supposed to work only with ```mate-session-manager```.

**Comment by monsta** [comment]
Ok, there goes https://github.com/mate-desktop/caja/pull/489, please test if it fixes the problem (and also doesn't introduce regressions when running in MATE).

**Comment by raveit65** [comment]
This fixes the issue for me, tested in f21 xfce VM.
....scratch build for testing is on the run.
thanks @monsta

**Comment by raveit65** [comment]
http://koji.fedoraproject.org/koji/taskinfo?taskID=12286978

**Comment by nirik** [comment]
Yep. Works fine here, thanks!

**Comment by monsta** [comment]
Cool. Merged https://github.com/mate-desktop/caja/pull/489.

**Comment by raveit65** [comment]
Issue is back with latest master from 2016-10-16.
@lukefromdc @monsta 
I suspect it is caused by GtkApplication commit.
I recall there was one commit where a check for several sm-manger was introduced, but i don't find this after the rebase.

**Comment by lukefromdc** [comment]
Try https://github.com/mate-desktop/caja/pull/654, which may solve this problem. Be sure to build with GtkApplication, then see if you can close Caja with one caja --quit or killall caja in XFCE. I don't have XFCE installed but this worked fine in a root caja session, where Caja immediately quit due to the exit on last window default, stayed shut down with no restart loop, and could be manually restarted with --force-desktop, which does not work if caja is already running.

**Comment by lukefromdc** [comment]
I just managed toinstall enough of XFCE to get a session running. I set up Caja to autostart, using today's master and managed to duplicate this issue. It was even worse when I manually set caja to autostart in XFCE's configuration. XFCE manages its own desktop and Caja defaulted to no desktop, but opened a window, then when I tried to close that window another one opened! 

Switching to a build with https://github.com/mate-desktop/caja/pull/654 solved this problem completely. Should I merge this? It only affects the GtkApplication builds and fixes this issue.
